### ISSUE TITLE
Incorrect type inferred when using spread operator to override objects with optional properties.

### ISSUE THREAD

**Issue by ioExpander** [microsoft/TypeScript#57408]
### 🔎 Search Terms

spread operator, spread operator optional, spread operator override

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.3#code/C4TwDgpgBA8gNgEygXigbygQwFxQM7ABOAlgHYDmAvgFCiRQByEA7gMrEC2Yc0qamAflwESFAD4BXUgggAzMhAQBuGtVlSAxsGIB7UlB0A3CIRIyAFDsS54CADRRSLdlx64mbTtwgBKbP2EiMioAArRqKEioQghgCUJ9NAA6FKt7KBSkp09XCEolalUNPTwrCCS4HXJLY1NiCwCoACJMJsoHRqkZeScESh8fagB6IbQWptwuuQU+oA

### 💻 Code

```ts
type Old = { a: string}
type NewSimple = {a?: string|undefined;}

function override(old: Old, newSimple: NewSimple):{a: string} {
    return {...old, ...newSimple};
}

console.log(override({a: "a"}, {a: undefined}))
// => {"a": undefined}
```

### 🙁 Actual behavior

This code compiles using all ts versions I have tried.

### 🙂 Expected behavior

I'm expecting the return type to be inferred as  `{a: string | undefined}`

### Additional information about the issue

This can lead to pretty nasty bugs as it won't catch some undefined values in the return type.

**Comment by MartinJohns** [comment]
This is working as intended and why `exactOptionalPropertyTypes` was added. Enable this flag and you get an error.

**Comment by RyanCavanaugh** [comment]
> In that case, why does ts removes the undefined after the spread operator if that undefined value is allowed?

I don't really understand the question -- you seem to be asking why TS acts like EOPT is off when EOPT is off.

**Comment by ioExpander** [comment]
> I don't really understand the question -- you seem to be asking why TS acts like EOPT is off when EOPT is off.

Yes. Most likely I'm reading the TSDoc wrong, so thank you for your help. Let me try again : 

As I read it, the behavior described in the TSDoc for EOPT off seems to be `{a?: string}` is "expanded" to `{a?: string|undefined}`. 

The TS doc does not explain why, with **EOPT off**,  "{...{a: string}, ...{a?: string|undefined}}" is inferred as `{a: string}` as if I had written "{...{a: string}, ...{a?: string}}" **with EOPT on**

With **EOPT on** I'm would have expected "{...{a: string}, ...{a?: string}}" to be inferred as {a: string}.
But with **EOPT off** I would have expected "{...{a: string}, ...{a?: string}}" to be expanded to "{...{a: string}, ...{a?: string|undefined}}" and both inferred as `{a: string| undefined}`

**Comment by RyanCavanaugh** [comment]
The read type on `{ a?: string }` is always `string | undefined`.

The correct mental model for EOPT is that if EOPT is off, you're saying that in your program, the runtime values `{ a: undefined }` and `{ }` have the same behavior. Obviously that's never true for spread, but the behavior in all cases is to assume that an `{ a?: string }` actually is always `{ }` (since the opposite assumption would be pretty unusable in practice). EOPT is the rule which enforces that that assumption is actually correct.

**Comment by ioExpander** [comment]
OK I get it now. Thanks a lot !

> The correct mental model for EOPT is that if EOPT is off, you're saying that in your program, the runtime values `{ a: undefined }` and `{ }` have the same behavior. Obviously that's never true for spread, but the behavior in all cases is to assume that an `{ a?: string }` actually is always `{ }` (since the opposite assumption would be pretty unusable in practice). EOPT is the rule which enforces that that assumption is actually correct.

⬆️ Your explanation ⬆️ is way better, IMHO, than the [current one](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) in TSDocs. Maybe adding this in the docs, including the potential gotcha for the spread/shallow merge, can help others anticipate such an issue.

I wish there was a way to fine tune EOPT to prevent TS from letting undefined values slip through in the specific case of spread syntax. For sure it would help catch potential bugs for careless uses of spread like my initial example and as such keep the safety net TS brings and improve DX. Not sure if this is doable via TS. Maybe an ESLint warning ?

Again thanks a bunch for your help yesterday. 🙏
